src/tests/tests.h

```diff
@@ -656,6 +784,58 @@ class Text_Based_Test : public Test
       std::vector<uint64_t> m_cpu_flags;
    };
 
+/**
+ * This is a convenience wrapper to write small self-contained and in particular
+ * exception-safe unit tests. If some (unexpected) exception is thrown in one of
+ * the CHECK-scopes, it will fail the particular test gracefully with a human-
+ * understandable failure output.
+ *
+ * Example Usage:
+ *
+ * ```
+ * std::vector<Test::Result> test_something()
+ *    {
+ *    return
+ *       {
+ *       CHECK("some unit test name", [](Test::Result& r)
+ *          {
+ *          r.confirm("some observation", 1+1 == 2);
+ *          }),
+ *       CHECK("some other unit test name", [](Test::Result& r)
+ *          {
+ *          // ...
+ *          })
+ *       };
+ *    }
+ *
+ * BOTAN_REGISTER_TEST_FN("some_category", "some_test_name", test_something);
+ * ```
+ */
+template<typename FunT>
+Test::Result CHECK(const char* name, FunT check_fun)
+   {
+   Botan_Tests::Test::Result r(name);
+
+   try
+      {
+      check_fun(r);
+      }
+   catch(const Botan_Tests::Test_Aborted&)
+      {
+      // pass, failure was already noted in the responsible `require`
+      }
+   catch(const std::exception& ex)
+      {
+      r.test_failure(std::string("failed with exception: ") + ex.what());
```

Comment:
The two-argument variant of `test_failure` is intended for the case of `test_failure("some message", e.what())`